### PR TITLE
ENG-307: Add differential tests to the StarkEx.Crypto.SDK

### PR DESCRIPTION
**Pull request by asynctomatic**
## Describe your changes
The recent bug in signature verification on the StarkEx.Crypto.SDK raised awareness for the need to add a testing suit comparing our implementation of cryptographic operations against another (battle tested) codebase.

This PR adds a differential testing campaign against [Starkware's Python implementation](https://github.com/starkware-libs/starkex-resources/blob/master/crypto/starkware/crypto/signature/signature.py) for the following methods:
- `PedersenHash:CreateHash`
- `StarkExSigner:SignMessage`
- `StarkExSigner:VerifySignature`
- `MathUtils:MimicEcMultiplicationAir`

## Issue ticket number and link
https://linear.app/threesigma/issue/ENG-307

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.

## Reviewers
@JoseAfonsoThreeSigma @asynctomatic

### CODE REVIEW

**Review by JoseAfonsoThreeSigma — APPROVED** [review]
LGTM, just a small improvement for the future: ideally we would checkout the python implementation or add that repo as a git-submodule, and test our implementation against the most recent one. Currently, we have to update this implementation manually and we are not subject to the updates that are done on that repo. If you agree, we can open this as a GH issue and proceed with this PR

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Signing/StarkExSignerDifferentialTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    [Theory]
+    [Repeat(10)]
+    public void VerifySignature_DifferentialTesting_ResultsMatch(int runId)
+    {
+        // Generate random inputs.
+        var messageHash = RandomHelpers.GetRandomBigInteger(new BigInteger("2").Pow(251)).ToString(16);
+        var privateKey = RandomHelpers.GetRandomBigInteger(new BigInteger("2").Pow(251)).ToString(16);
+
+        // Compute other inputs.
+        var target = CreateStarkExSigner();
```
Comment (3s-jorgematos) [inline tests/StarkEx.Crypto.SDK.DifferentialTests/Signing/StarkExSignerDifferentialTests.cs:41]: Why is this needed when target is not used and the method is static? Can this be deleted?

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+namespace StarkEx.Crypto.SDK.DifferentialTests.Utils;
```
Comment (3s-jorgematos) [inline tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs:1]: I see some warnings here on GitHub like:

`The file header is missing or not located at the top of the file.`  & `Prefix local calls with this` but they don't show locally not sure why. Any idea?

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Helpers/RandomHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+namespace StarkEx.Crypto.SDK.DifferentialTests.Helpers;
+
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Security;
+
+/// <summary>
+/// A collection of helper methods for generating random values.
+/// </summary>
+public class RandomHelpers
```
Comment (3s-jorgematos) [inline tests/StarkEx.Crypto.SDK.DifferentialTests/Helpers/RandomHelpers.cs:9]: This class can be static

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Helpers/PythonHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+using System.Diagnostics;
+using System.Text;
+using Nethereum.Hex.HexConvertors.Extensions;
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Math.EC;
+using StarkEx.Commons.SDK.Models;
+using StarkEx.Crypto.SDK.Signing;
+
+public class PythonHelpers
```
Comment (3s-jorgematos) [inline tests/StarkEx.Crypto.SDK.DifferentialTests/Helpers/PythonHelpers.cs:11]: This class can also be static

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        var result = target.CreateHash(leftField, rightField);
+
+        // Run Python implementation.
+        var controlResult = PythonHelpers.Hash(leftField, rightField);
+
+        // Compare results.
+        result.Should().Be(controlResult);
+    }
+
+    private IPedersenHash CreatePedersenHash()
```
Comment (3s-jorgematos) [inline tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs:33]: This method can be static too

**Review by 3s-jorgematos — CHANGES_REQUESTED** [review]
Just some small changes

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+using StarkEx.Crypto.SDK.Hashing;
+using StarkEx.Crypto.SDK.Signing;
+using Xunit;
+
+[ExcludeFromCodeCoverage]
+public class PedersenHashDifferentialTests
+{
+    [Theory]
+    [Repeat(10)]
+    public void CreateHash_InputsAreValid_ResultsMatch(int runId)
```
[the remark this answers is not in the document]
Comment (3s-jorgematos) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs:16]: After checking this again, the value comes from the Repeat attribute. Never seen this pattern, can you share some reference?

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Signing/StarkExSignerDifferentialTests.cs (hunk 1 above)
Comment (asynctomatic) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Signing/StarkExSignerDifferentialTests.cs:41]: I'm not sure of I follow here, what method are you referring to?

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (asynctomatic) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs:16]: This is a weird pattern that I found online to repeat a XUnit test an X amount of times. Without the `runId` parameter, XUnit considers the tests runs to be the same and skips them.

I found this here:
https://stackoverflow.com/a/55687930/20987707

How do you think we should proceed here? I can try and change it to use the `MemberData` attribute.

**Comment by asynctomatic** [comment]
@JoseAfonsoThreeSigma Your approach makes sense @JoseAfonsoThreeSigma even thought would probably always need to manually update our python SDK wrapper that is called in the differential tests. I think the approach here if using a submodule would be to lock a specific commit and manually update it when needed (I don't expect this core functionality in their crypto SDK to change much thought).

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs (hunk 2 above)
Comment (asynctomatic) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs:1]: Not sure, this is weird. Do you have any input here @JoseAfonsoThreeSigma ?

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Signing/StarkExSignerDifferentialTests.cs (hunk 1 above)
Comment (3s-jorgematos) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Signing/StarkExSignerDifferentialTests.cs:41]: The variable `target` has no uses, and since the method `CreateStarkExSigner` is static, there are no side effects besides the output (and eventually on the input for reference variables, but it is not the case here)

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (3s-jorgematos) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs:16]: Its fine by me like this. I would just disable it with [pragma ](https://learn.microsoft.com/en-us/cpp/preprocessor/warning?view=msvc-170) or something similar

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs (hunk 2 above)
Comment (JoseAfonsoThreeSigma) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs:1]: Not really, I also don't have these warnings locally. We could try to add the proper rule to disable this warning to our `.editorconfig`. https://github.com/dotnet/docs/blob/main/docs/fundamentals/code-analysis/style-rules/ide0073.md

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (asynctomatic) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Hashing/PedersenHashDifferentialTests.cs:17]: Disabled with pragma

File: tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs (hunk 2 above)
Comment (asynctomatic) [reply tests/StarkEx.Crypto.SDK.DifferentialTests/Utils/MathUtilsDifferentialTests.cs:1]: Added `dotnet_diagnostic.IDE0073.severity = none` to the `.editorconfig` file

**Review by 3s-jorgematos — APPROVED** [review]
LGTM
